Add RoleHierarchyBuilder

FdHerrera opened the pull request:
Create builder class to declaratively create a RoleHierarchy with a fluent API, add test. Closes #13300
gh-13300

jzheaux commented on core/src/main/java/org/springframework/security/access/hierarchicalroles/RoleHierarchyBuilder.java:
```diff
@@ -0,0 +1,68 @@
+		this.roleBranches = new LinkedHashMap<>();
+	}
+
+	public static RoleHierarchyBuilder newBuilder() {
```
Since any role hierarchy will require at least one role, I think it would be nice to call this `role`, removing one call.

jzheaux replied:
Note that I believe it is moved into `RoleHierarchyImpl`, then the two `role` methods won't conflict since this one will live in `RoleHierarchyImpl` itself.

FdHerrera replied:
Hello @jzheaux 

While implementing this suggestion I think placing the call to the builder in the implementation is kind of mixing the roles of the entities.

I suggest, instead, leave the building of the `RoleHierarchyImpl` to live only in the builder, and refactor the final `build()` method to create a fully featured `RoleHierarchyImpl` (using the utility method `RoleHierarchyUtills#roleHierarchyFromMap` and `RoleHierarchyImpl#setHierarchy`, this way, it will have a complete and clearer scope and will reduce the impact on related classes, it will also allow doing something like the follow:

```
@Bean
public RoleHierarchy roleHierarchy() {
    return RoleHierarchyBuilder.newBuilder()
        .role("ROLE_A").impliesAllOf("ROLE_B")
        .role("ROLE_B").impliesAllOf("ROLE_C", "ROLE_D")
        .build();
}
```

Elegant and concise, really straightforward. 
Wdyt?

FdHerrera replied:
Thank you for your feedback and help to further understand your idea, @jzheaux !

Makes total sense so I'm planning to move the builder to the class `RoleHierarchyImpl` to follow the convention of the classes you mention.

Will try to implement  a `role(String)` method in that class to better align with the last comment in this thread.

Thank you for sharing your thoughts!

jzheaux commented on core/src/main/java/org/springframework/security/access/hierarchicalroles/RoleHierarchyBuilder.java:
```diff
@@ -0,0 +1,68 @@
+		 * @throws IllegalArgumentException if <code>impliedRoles</code> is null or empty
+		 * or contains any null element.
+		 */
+		public RoleHierarchyBuilder inheritsFromAllOf(String... impliedRoles) {
```
Can we call this `implies`? I think the simpler name makes the DSL more readable. It also aligns the method name with the parameter name and JavaDoc.

jzheaux commented on core/src/main/java/org/springframework/security/access/hierarchicalroles/RoleHierarchyBuilder.java:
```diff
@@ -0,0 +1,68 @@
+ * Class to provide a fluent API to build a {@link String} representing a {@link RoleHierarchy}.
+ * Relies on  {@link RoleHierarchyUtils#roleHierarchyFromMap(Map)} to build the string.
+ */
+public class RoleHierarchyBuilder {
```
Please make the class `final` until it's clear that extendability is needed.

jzheaux commented on core/src/main/java/org/springframework/security/access/hierarchicalroles/RoleHierarchyBuilder.java:
```diff
@@ -0,0 +1,68 @@
+
+/**
+ * Class to provide a fluent API to build a {@link String} representing a {@link RoleHierarchy}.
+ * Relies on  {@link RoleHierarchyUtils#roleHierarchyFromMap(Map)} to build the string.
```
Please add `@since 6.3`. Please also consider adding an `@author` tag.

jzheaux commented on core/src/main/java/org/springframework/security/access/hierarchicalroles/RoleHierarchyBuilder.java:
```diff
@@ -0,0 +1,68 @@
+		return roleHierarchyFromMap(roleBranches);
+	}
+
+	public static class RoleBranchBuilder {
```
Please also make this class `final`.

jzheaux commented on core/src/main/java/org/springframework/security/access/hierarchicalroles/RoleHierarchyBuilder.java:
```diff
@@ -0,0 +1,68 @@
+		private final RoleHierarchyBuilder parentBuilder;
+		private final String role;
+
+		RoleBranchBuilder(String role, RoleHierarchyBuilder roleHierarchyBuilder) {
```
Please make the constructor `private`.

jzheaux reviewed (changes requested):
Thanks for the PR, @FdHerrera! Given that this is ultimately in the service of `RoleHierarchyImpl`, will you please move this into that class? In that way, I believe it will also better align with https://github.com/spring-projects/spring-security/pull/13788

FdHerrera commented:
Thank you very much for your feedback, @jzheaux ! 

I'll start implementing your suggestions.

jzheaux commented on core/src/main/java/org/springframework/security/access/hierarchicalroles/RoleHierarchyImpl.java:
```diff
@@ -213,4 +222,65 @@ else if (roleName.equals(lowerRole.getAuthority())) {
+			 * @throws IllegalArgumentException if <code>impliedRoles</code> is null or
+			 * empty or contains any null element.
+			 */
+			public Builder impliesAllOf(String... impliedRoles) {
```
I think it would be better if this were called `implies`. The reason is that `allOf` is implied by the fact that it is a varargs.

I think the `allOf` prefix makes sense for things like `AuthorizationManagers` since you can have an `anyOf` there too, which makes the clarification necessary in that case.

jzheaux replied:
Looking back at my first review, I realize that one of my comments used `implies` and another used `impliesAllOf`. Sorry for potentially being confusing about my position.

jzheaux commented on core/src/main/java/org/springframework/security/access/hierarchicalroles/RoleHierarchyImpl.java:
```diff
@@ -99,6 +104,10 @@ public class RoleHierarchyImpl implements RoleHierarchy {
 	 */
 	private Map<String, Set<GrantedAuthority>> rolesReachableInOneOrMoreStepsMap = null;
 
+	public static Builder.RoleBranchBuilder role(String role) {
```
I believe [this comment](https://github.com/spring-projects/spring-security/issues/13300#issuecomment-1606115443) raises something that this PR hasn't considered yet, which is role prefixes.

What if instead we went back to your earlier suggestion and did the following:

```java
public static Builder builder() {
    return new Builder("ROLE_");
}

public static Builder withRolePrefix(String rolePrefix) {
    return new Builder(rolePrefix);
}
```

The reason is that, in Spring Security, things like `hasRole("USER")` imply that `ROLE_` will be prefixed automatically, as opposed to `hasAuthority(`ROLE_USER`)` where the application must supply the prefix. `role` methods in this class should do the same thing.

I think this arrangement of methods is better aligned since they both return the same `Builder` class and we also don't introduce a `role` method that doesn't automatically prefix the underlying authority.

FdHerrera replied:
Sure, makes sense, `implies` should also prefix ROLE to each one? Something like
```
role("FOO").implies("BAR")
```

resulting in: `ROLE_FOO > ROLE_BAR` ?

FdHerrera replied:
Or whatever prefix is defined using `withRolePrefix`

I'm thinking maybe `builderWithRolePrefix` ? Or just overload `builder` with a parameter `rolePrefix`?

Just because `RoleHierarchyImpl.withRolePrefix(rolePrefix)` doesn't sound very intuitive to me. Wdyt?

jzheaux replied:
> `role("FOO").implies("BAR")`

Yes, that's the idea.

> Or whatever prefix is defined using `withRolePrefix`

Good question. The `with` prefix comes from a naming convention in Spring Security. `withXXX` means "give me a builder based on XXX" and `fromXXX` means "construct an instance using values XXX". An example is `NimbusJwtDecoder#withJwkSetUri`.

If what you are looking for is to align the method names, maybe we consider something like `withDefaults()` or `withDefaultRolePrefix()` instead of `builder()`. I kind of like the subtle reminder that `withDefaultRolePrefix()` will include the default role prefix for you.

FdHerrera replied:
Liked the `withDefaultRolePrefix()`

Went ahead with this approach and implemented another one with no prefix, for cases where no standard prefix is used.

That led to 3 factory methods to build a hierarchy: `withDefaultRolePrefix`, `withRolePrefix` and `withNoRolePrefix`

Let me know if it looks good for you! Ty in advance

jzheaux commented on core/src/main/java/org/springframework/security/access/hierarchicalroles/RoleHierarchyImpl.java:
```diff
@@ -213,4 +222,65 @@ else if (roleName.equals(lowerRole.getAuthority())) {
+			this.roleBranches = new LinkedHashMap<>();
+		}
+
+		public RoleBranchBuilder role(String role) {
```
Thanks for adding JavaDoc. Will you please add a JavaDoc here, too?

jzheaux reviewed (changes requested):
Thanks, @FdHerrera! Things are shaping up nicely. I've left a little bit more feedback inline.

Also, in preparation for merging, will you please squash your commits into a single commit that looks something like this:

```
Add RoleHierarchy.Builder

Closes gh-13300
```

FdHerrera commented:
Thank you for the thorough review and constructive feedback! I appreciate the collaborative back-and-forth @jzheaux .

I implemented your suggestions and squashed the commits, I think is looking pretty good let me hear your thoughts!

FdHerrera commented:
> I believe it will also better align with #13788

Also, regarding this comment, should I change the commit and description to close that issue instead of gh-13300?

jzheaux commented:
Thanks, @FdHerrera! I've added a polish to make it align more closely with the rest of the codebase stylistically.

For now, I think we should leave out `withNoRolePrefix` and wait to see how the community uses the class. In other places in Spring Security, `ROLE_` is the default and you supply a string value to override the default. There isn't the notion elsewhere in the code base of "no prefix". I went ahead and adjusted that in my polish commit. Let me know if me removing that method causes a problem.

FdHerrera commented:
Looks good @jzheaux thank you for your help!
